lib/cdo/i18n_string_url_tracker.rb

```diff
@@ -76,18 +76,13 @@ def log(string_key, url, source)
 
   # Records the log data to a buffer which will eventually be flushed
   def add_to_buffer(string_key, url, source)
-    data = {url => {string_key => Set[source]}}
     # make sure this is the only thread modifying @buffer
     @buffer.synchronize do
-      # update the buffer size if we are adding any new data to it
-      # duplicate data will not increase the buffer size
-      size = 0
-      size += url.bytesize unless @buffer.dig(url)
-      size += string_key.bytesize unless @buffer.dig(url, string_key)
-      size += source.bytesize unless @buffer.dig(url, string_key)&.include?(source)
-      @buffer_size += size
       # add the new data to the buffer
-      @buffer.deep_merge!(data)
+      ((@buffer[url] ||= {})[string_key] ||= Set.new).add?(source) &&
+        # update the buffer size if we are adding any new data to it
+        # duplicate data will not increase the buffer size
+        @buffer_size += (url.bytesize + string_key.bytesize + source.bytesize)
```

Comment:
I don't think this estimation is correct because it will double count `url` and `string_key`. The previous code only increased the estimated size by :
* url.bytesize if this is the first time the url is being added
* string_key.bytesize if this is the first time the url + string_key pair is being added
* source.bytesize if this is the first time the url + string_key + source pair is being added.

The code in this diff will repeatedly add `url.bytesize` even if its a URL we already have in the hash.

Ultimately though, we weren't hitting these size limits when we last when to production, so maybe this size calculation isn't needed at all?

Reply:
Ah, I assumed this was estimating the size of the records eventually flushed to Firehose, but it's instead estimating the size of the internal Hash object. I adjusted the behavior to match what it was calculating previously, so whether it's needed or not it won't be changed by this.